### PR TITLE
Implementing PCEN and providing option to switch

### PR DESCRIPTION
**Pull request by ljj7975**

### CODE REVIEW

**Comment by lintool** [comment]
Do we have evaluation results and comparisons to existing models?

**Comment by ljj7975** [comment]
I have compared accuracy but I did not see any difference at least with our current resnets.

I have also tried PCEN provided by librosa but it was actually worse.
We decided to keep MFCCs for now and compare again with convrnn once it is implemented.

**Comment by ljj7975** [comment]
One possible reason might be that google data have similar loudness.
in the paper, they augmented their data by scaling the loudness to see the benefit of PCEN.
I will see if changing loudness pose any accuracy difference when I have time

File: utils/manage_audio.py [hunk 1]
```diff
@@ -7,19 +7,44 @@
...
+class AudioPreprocessor(object):
+    def __init__(self, sr=16000, n_dct_filters=40, n_mels=40, f_max=4000, f_min=20, n_fft=480, hop_ms=10):
+        super().__init__()
+        self.n_mels = n_mels
+        self.dct_filters = librosa.filters.dct(n_dct_filters, n_mels)
+        self.sr = sr
+        self.f_max = f_max if f_max is not None else sr // 2
+        self.f_min = f_min
+        self.n_fft = n_fft
+        self.hop_length = sr //1000 * hop_ms
```
Comment (daemon) [inline utils/manage_audio.py]: `//` missing space

File: utils/manage_audio.py [hunk 2]
```diff
@@ -7,19 +7,44 @@
...
+        self.pcen_transform = pcen.StreamingPCENTransform(n_mels=n_mels, n_fft=n_fft, hop_length=self.hop_length, trainable=True)
+
+    def get_MFCCs(self, data):
```
Comment (daemon) [inline utils/manage_audio.py]: I think `compute_mfccs` is a better name.

File: utils/manage_audio.py [hunk 3]
```diff
@@ -7,19 +7,44 @@
...
+            hop_length=self.hop_length,
+            n_fft=self.n_fft,
+            fmin=self.f_min,
+            fmax=self.f_max)
+        data[data > 0] = np.log(data[data > 0])
+        data = [np.matmul(self.dct_filters, x) for x in np.split(data, data.shape[1], axis=1)]
+        data = np.array(data, order="F").reshape(1, 101, 40).astype(np.float32)
+        return data
+
+    def get_PCEN(self, data):
```
Comment (daemon) [inline utils/manage_audio.py]: `compute_pcen`

File: README.md [hunk 4]
```diff
@@ -77,6 +77,7 @@ There are command options available:
 
 | option         | input format | default | description |
 |----------------|--------------|---------|-------------|
+| `--audio_preprocess_type`   | {MFCCs, PCEN}       | PCEN     | type of audio preprocess to use            |
```
Comment (daemon) [inline README.md]: If the new default is PCEN, then we'll need to update the existing instructions to use MFCCs.

**Review by daemon — CHANGES_REQUESTED** [review]
Overall structure looks good. A few minor issues and nitpicks here and there.

File: README.md (hunk 4 above)
Comment (ljj7975) [reply README.md]: oh ya I updated the code but forgot to update readme

**Review by daemon — APPROVED** [review]

**Comment by lydiaji** [comment]
> One possible reason might be that google data have similar loudness.
> in the paper, they augmented their data by scaling the loudness to see the benefit of PCEN.
> I will see if changing loudness pose any accuracy difference when I have time

I've also used PCEN, but the result is not as good as I simply use MFCC as features. In the paper, the results with or without multi loudness training makes no much difference. So I wonder why using PCEN does not improve my result? Would you please give me some suggestion? Thanks!

**Comment by daemon** [comment]
My guess is that you'd have to see more loudness variability in your audio data for PCEN to be useful, as is the case for more extreme near- and far-field settings, e.g., person talking 1cm away from mic vs. 5ft. Also, in the original paper they compare only to log-Mels, not MFCCs -- whether this makes a difference for deep neural nets is unclear, since MFCCs are just a linear map over log-Mels.

**Comment by lydiaji** [comment]
> My guess is that you'd have to see more loudness variability in your audio data for PCEN to be useful, as is the case for more extreme near- and far-field settings, e.g., person talking 1cm away from mic vs. 5ft. Also, in the original paper they compare only to log-Mels, not MFCCs -- whether this makes a difference for deep neural nets is unclear, since MFCCs are just a linear map over log-Mels.

Thanks for your reply. I have just noticed that I didn't use multi-loudness training. In the paper, for fixed PCEN, its accuracy with or without multi-loudness training performs similar. While for trainable PCEN, it needs multi-loudness training. So would it be a reason that the result is not so good?

**Comment by lydiaji** [comment]
In addition, I didn't use streaming data. Would it be a problem for computing iir filter? In the paper, it showed that 23 left context frames and 9 right context frames as input to the net. I'm not clear about how to implement that? Would you please give me some guide? Thanks!!!
